### PR TITLE
RD2: Schedule Service Handles DayOfMonth 29 and 30

### PR DESCRIPTION
**Pull request by npsp-reedestockton**
We've added explicit support for monthly recurring donation days 29 and 30 (via Day of Month field).  Previously, these days were all considered the last day of the month.  Now we will respect the explicit selection in all months where that is possible and treat the selection as the last day of the month in all other cases.  Last Day of Month is still supported as an option.

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by npsp-reedestockton** [comment]
@force2b @RadSFDeveloper Ready for review.
@sam-knox Please take a look at the PR description.

Thanks all!

**Comment by npsp-reedestockton** [comment]
***lurch: attach W-040426*

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001Hu46QAC" target="_blank">W-040426</a>

File: src/classes/RD2_ScheduleService.cls [hunk 1]
```diff
@@ -477,6 +477,7 @@ public without sharing class RD2_ScheduleService {
             adjusted = addDateUnits(adjusted, adjustedFrequency, schedule.InstallmentPeriod__c);
         }
 
+        //TODO: Shouldn't this test for any day > days of month??
```
Comment (force2b) [inline src/classes/RD2_ScheduleService.cls]: What do you mean by this?

File: src/classes/RD2_ScheduleService_TEST.cls [hunk 2]
```diff
@@ -258,6 +258,63 @@ private with sharing class RD2_ScheduleService_TEST {
...
+        RD2_ScheduleService service = getScheduleServiceForCurrentDate(today);
+
+        System.assertEquals(nextDonationDate, service.getNextDonationDate(rd), 'Next Donation Date should match');
+    }
+
+    /***
+    * @description Verifies next donation date for monthly donation when donation falls in next month.
+    */
+    @isTest
+    private static void shouldGenerateNextDonationDateLastDayInFebruaryWhenDayIs31() {
```
Comment (force2b) [inline src/classes/RD2_ScheduleService_TEST.cls]: Why is this needed?

File: src/classes/RD2_ValidationService.cls [hunk 3]
```diff
@@ -285,7 +285,7 @@ public with sharing class RD2_ValidationService {
             } catch (Exception e) {
             }
 
-            Boolean isValid = dayOfMonth > 0 && dayOfMonth < 29;
+            Boolean isValid = dayOfMonth > 0 && dayOfMonth < 32;
```
Comment (force2b) [inline src/classes/RD2_ValidationService.cls]: This should be `< 31`

File: src/classes/RD2_ValidationService_TEST.cls [hunk 4]
```diff
@@ -314,7 +314,7 @@ private with sharing class RD2_ValidationService_TEST {
             .withContact(getContact().Id);
 
         for (String day : new Set<String>{
-            '29', '30', '31'
+            '29', '30', '31', RD2_Constants.DAY_OF_MONTH_LAST_DAY
```
Comment (force2b) [inline src/classes/RD2_ValidationService_TEST.cls]: This should not include `31`

File: src/objects/RecurringDonationSchedule__c.object [hunk 5]
```diff
@@ -333,6 +333,21 @@
...
+                    <fullName>29</fullName>
+                    <default>false</default>
+                    <label>29</label>
+                </value>
+                <value>
+                    <fullName>30</fullName>
+                    <default>false</default>
+                    <label>30</label>
+                </value>
+                <value>
```
Comment (force2b) [inline src/objects/RecurringDonationSchedule__c.object]: This should not include `31`

File: src/objects/npe03__Recurring_Donation__c.object [hunk 6]
```diff
@@ -257,6 +257,11 @@
                     <default>false</default>
                     <label>31</label>
                 </value>
+                <value>
```
Comment (force2b) [inline src/objects/npe03__Recurring_Donation__c.object]: This needs to be removed. The "Last Day of Month" picklist is not packaged.

File: src/classes/RD2_ValidationService.cls (hunk 3 above)
Comment (RadSFDeveloper) [reply src/classes/RD2_ValidationService.cls]: @npsp-reedestockton , see how PR #5286 is handling validation service and other similar common changes, so you can just copy over common changes required for the of both PRs. Of course, feel free to add new tests if required.

**Comment by npsp-reedestockton** [comment]
@force2b several of your comments relate to DayOfMonth == 31.  Easy to remove.  I put it in for discussion around what we expect to be coming from Elevate.  They will never send us 31, right?

**Comment by force2b** [comment]
31 does not need to be supported _here_.

**Comment by andrewyu-salesforce** [comment]
@npsp-reedestockton I think you attach this PR to the wrong WI. It should be attach to WI-040427 instead of WI-040426.

File: src/classes/RD2_ValidationService_TEST.cls (hunk 4 above)
Comment (force2b) [reply src/classes/RD2_ValidationService_TEST.cls]: I'm pretty sure that 31 should not be here.

**Comment by npsp-reedestockton** [comment]
**lurch: attach W-040427

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001Hu4BQAS" target="_blank">W-040427</a>

**Comment by npsp-reedestockton** [comment]
**lurch: detach

**Comment by npsp-reedestockton** [comment]
**lurch: attach W-040427

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001Hu4BQAS" target="_blank">W-040427</a>

**Comment by LurchTheButler** [comment]
Detached <a href="https://foundation.my.salesforce.com/a2x1E000001Hu46QAC" target="_blank">W-040426</a>

File: src/classes/RD2_ScheduleService.cls [hunk 7]
```diff
@@ -480,6 +480,9 @@ public without sharing class RD2_ScheduleService {
         if (schedule.DayOfMonth__c == RD2_Constants.DAY_OF_MONTH_LAST_DAY && schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY) {
             adjusted = Date.newInstance(adjusted.year(), adjusted.month(), Date.daysInMonth(adjusted.year(),adjusted.month()));
         }
+        else if (Schedule.DayOfMonth__c != null && Integer.valueOf(schedule.DayOfMonth__c) > Date.daysInMonth(firstValid.year(), firstValid.month())) {
```
Comment (force2b) [inline src/classes/RD2_ScheduleService.cls]: Beside moving the else up to the line with the bracket, I think the operator on this incorrect. If `DayofMonth__c` is more than the number of days in that month, then creating a date with that value should fail. I think the operation should be less-than-or-equal.

```suggestion
        } else if (schedule.DayOfMonth__c != null && Integer.valueOf(schedule.DayOfMonth__c) <= Date.daysInMonth(firstValid.year(), firstValid.month())) {
```

Comment (npsp-reedestockton) [reply src/classes/RD2_ScheduleService.cls]: Comment is outdated.

Comment (npsp-reedestockton) [reply src/classes/RD2_ScheduleService.cls]: Refactored that code, so please review newest.  Thanks!

File: src/classes/RD2_ScheduleService.cls [hunk 8]
```diff
@@ -477,8 +477,12 @@ public without sharing class RD2_ScheduleService {
             adjusted = addDateUnits(adjusted, adjustedFrequency, schedule.InstallmentPeriod__c);
         }
 
-        if (schedule.DayOfMonth__c == RD2_Constants.DAY_OF_MONTH_LAST_DAY && schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY) {
-            adjusted = Date.newInstance(adjusted.year(), adjusted.month(), Date.daysInMonth(adjusted.year(),adjusted.month()));
+        if (schedule.InstallmentPeriod__c == RD2_Constants.INSTALLMENT_PERIOD_MONTHLY) {
+            Integer nextDayOfMonth =
+                schedule.DayOfMonth__c == RD2_Constants.DAY_OF_MONTH_LAST_DAY ?
+                    Date.daysInMonth(adjusted.year(),adjusted.month()) :
+                    Integer.valueOf(schedule.DayOfMonth__c);
```
Comment (force2b) [inline src/classes/RD2_ScheduleService.cls]: @npsp-reedestockton Does this still work if the `DayOfMonth__c` field value is 30 and the month is 2?

**Comment by npsp-reedestockton** [comment]
Ready for another look...  Thanks!

File: src/classes/RD2_ScheduleService_TEST.cls [hunk 9]
```diff
@@ -894,13 +949,214 @@ private with sharing class RD2_ScheduleService_TEST {
...
-            'Final installment date should be 10/1/2020.');
+        System.assertEquals(nextDonationDate, installments[ expectedNumber - 1].nextDonationDate,
+                'Final installment date should be 10/1/2020.');
+    }
+
+    /***
+    * @description Verifies visualization works as expected across short month with day of month > days in short month
+    */
+    @isTest
+    private static void shouldReturnThreeInstallmentsSpanningFebruaryInNormalYear() {
```
Comment (RadSFDeveloper) [inline src/classes/RD2_ScheduleService_TEST.cls]: These four tests are verifying expected Next Donation Dates, so we can replace number of returned installments since that is arbitrary with changed tests name, for example:
`shouldReturnThreeInstallmentsSpanningFebruaryInNormalYear() ` -> `shouldReturnExpectedNextDonationDatesSpanningFebruaryInNormalYear()`?

File: src/classes/RD2_ScheduleService_TEST.cls [hunk 10]
```diff
@@ -894,13 +949,214 @@ private with sharing class RD2_ScheduleService_TEST {
...
+
+        npe03__Recurring_Donation__c rd = getRecurringDonationMonthlyBuilder()
+                .withDayOfMonth(dayOfMonth)
+                .build();
+
+        Test.startTest();
+        insert rd;
+        Test.stopTest();
+
+        List<RecurringDonationSchedule__c> schedules = [
```
[the remark this answers is not in the document]
Comment (npsp-reedestockton) [reply src/classes/RD2_ScheduleService_TEST.cls:980]: Thank you @RadSFDeveloper.  That's much better.  Will modify those four tests as you suggest.
